### PR TITLE
sync paras in program, fix deepcopy bug, python3 compatibility

### LINKED ISSUE #22: [bug] Here should check val[k] whether is a Network or LayerFunc
https://github.com/PaddlePaddle/PARL/blob/e48f67a644987d3ecca8c1d09da5d8d740563bdc/parl/layers/layer_wrappers.py#L156

### PR DESCRIPTION
**Pull request by zenghsh3**
- resolve #22 
- sync parameters in program run (won't copy data from gpu to cpu)
- fix bug of  deepcopy in Network
- add get_parameter_names api of Network
- python3 compatibility

### CODE REVIEW

File: parl/framework/model_base.py [hunk 1]
```diff
@@ -15,47 +15,115 @@
...
 
+import paddle.fluid as fluid
 from abc import ABCMeta, abstractmethod
 from parl.utils.utils import has_func
+from parl.layers.layer_wrappers import LayerFunc
 
 __all__ = ['Network', 'Model']
 
 
+def _fetch_framework_var(attr_name, is_bias):
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/model_base.py]: Comments should be added here.

File: parl/framework/model_base.py [hunk 2]
```diff
@@ -15,47 +15,115 @@
...
+    shape = core_var.get_tensor().shape()
+    framework_var = fluid.layers.create_parameter(
+        shape=shape,
+        dtype='float32',
+        attr=fluid.ParamAttr(name=attr_name),
+        is_bias=is_bias)
+    return framework_var
+
+
+def _gen_paras_pair(src, target):
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/model_base.py]: Explain why we need this function or where this function could be used for.

File: parl/framework/model_base.py [hunk 3]
```diff
@@ -15,47 +15,115 @@
...
+        for k in src.keys():
+            assert k in target
+            paras_pair.extend(_gen_paras_pair(src[k], target[k]))
+    else:
+        # for any other type, won't be handled
+        pass
+    return paras_pair
+
+
+def _get_parameter_names(obj):
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/model_base.py]: Comments should be added.

File: parl/framework/model_base.py (hunk 1 above)
Comment (zenghsh3) [reply parl/framework/model_base.py]: done

File: parl/framework/model_base.py (hunk 2 above)
Comment (zenghsh3) [reply parl/framework/model_base.py]: done

File: parl/framework/model_base.py (hunk 3 above)
Comment (zenghsh3) [reply parl/framework/model_base.py]: done

File: parl/framework/tests/model_base_test.py [hunk 4]
```diff
@@ -12,31 +12,405 @@
...
+                pred_program, feed={'obs': x}, fetch_list=[model_output])
+
+            outputs_2 = self.executor.run(
+                pred_program_2,
+                feed={'obs': x},
+                fetch_list=[target_model_output])
+            self.assertEqual(outputs[0].flatten(), outputs_2[0].flatten())
+
+    def _numpy_update(self, target_model, decay):
+        model_fc1_w = np.array(global_scope().find_var('fc1.w').get_tensor())
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/tests/model_base_test.py]: fetching value of a specific variable is a quite common function in paddle, could we provide this function in PARL?

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM:)

**Pull request merged**